criu/pages-compress.c

```diff
@@ -0,0 +1,208 @@
+	char pathsrc[PATH_MAX];
+	char pathdst[PATH_MAX];
+	const long pid = getpid();
+	const int dfd = get_service_fd(IMG_FD_OFF);
+	const int id = 1; // the only image we'd like to compress
+	int ret;
+
+	LZ4IO_prefs_t *lz4_prefs = LZ4IO_defaultPreferences();
+	LZ4IO_setBlockSize(lz4_prefs, 128 * 1024);
+
+	pr_debug("Compressing pages\n");
+
+	sprintf(format, "%s/%s", "/proc/%ld/fd/%d", imgset_template[CR_FD_PAGES].fmt);
+	sprintf(pathsrc, format, pid, dfd, id);
+
+	sprintf(format, "%s/%s", "/proc/%ld/fd/%d", imgset_template[CR_FD_PAGES_COMP].fmt);
+	sprintf(pathdst, format, pid, dfd, id);
+
+	ret = LZ4IO_compressFilename(pathsrc, pathdst, 8, lz4_prefs);
+	LZ4IO_freePreferences(lz4_prefs);
+
+	if (ret) {
+		pr_err("Can't compress %s to %s\n", pathsrc, pathdst);
+		return -1;
+	}
+	if (unlink(pathsrc)) {
+		pr_perror("Can't delete uncompressed source pages image: %s\n", pathsrc);
+		return -1;
+	}
+	return 0;
+}
+
+static int decompress_image(int comp_fd) {
+
+	const int compbufsize = 64 * 1024;
+	char compbuf[compbufsize];
+	const int outbufsize = 4 * 1024 * 1024; // LZ4F_max4MB is the max block size; see LZ4F_blockSizeID_t for details
+	char outbuf[outbufsize];
+	LZ4F_errorCode_t lz4err;
+	LZ4F_decompressionContext_t dctx;
+	int ret;
+	size_t totalread = 0;
+	size_t totalwrite = 0;
+	int bytestoread = compbufsize;
+	size_t offset = 0;
+
+	lz4err = LZ4F_createDecompressionContext(&dctx, LZ4F_VERSION);
+	if (LZ4F_isError(lz4err)) {
+		pr_err("Can't create LZ4 decompression context\n");
+		return -1;
+	}
+
+	pr_debug("Creating file at '%s'\n", temp_path_decompressed);
+	ret = open(temp_path_decompressed, O_CREAT | O_WRONLY | O_TRUNC, 0600);
+
+	if (0 > ret) {
+		pr_err("Can't create file, errno=%d\n", errno);
+		LZ4F_freeDecompressionContext(dctx);
+		return -1;
+	}
+
+	while (1) {
+		const size_t readbytes = read(comp_fd, compbuf + offset, bytestoread);
+		totalread += readbytes;
+		if (!readbytes && 0 == offset) {
+			/* reached end of file or stream */
+			break;
+		}
+		{
+			size_t outsize = sizeof outbuf;
+			size_t insize = offset + readbytes;
+			const size_t insize_orig = insize;
+			lz4err = LZ4F_decompress(dctx, outbuf, &outsize, compbuf, &insize, NULL);
+			if (LZ4F_isError(lz4err)) {
+				pr_err("LZ4 Decompress error: %s\n", LZ4F_getErrorName(lz4err));
+				break;
+			}
+			{
+				const ssize_t res = write(ret, outbuf, outsize);
+				if (0 > res) {
+					pr_err("write error to output file, fd=%d, errno=%d\n", ret, errno);
+					break;
+				}
+			}
+			totalwrite += outsize;
+			offset = insize_orig - insize;
+			bytestoread = insize;
+			memmove(compbuf, compbuf + compbufsize - offset, offset);
+		}
+	}
+	close(ret);
+	pr_debug("decompression completed, read %lu, wrote %lu\n", (unsigned long)totalread, (unsigned long)totalwrite);
+	return 0;
+}
+
+static pthread_t decomp_thread = 0;
+
+static void *decompression_thread_routine(void *param) {
+	int ret;
+	int comp_fd = (intptr_t)param;
+	ret = decompress_image(comp_fd);
+	close(comp_fd);
```

Comment:
Why not to open a compressed file descriptor also in this thread? 
open() and close() will be in the same thread in this case.

Reply:
Opening a compressed file in 'decompression_thread_start' is required to automatically detect if decompression needs to be enabled on 'decompression_thread_start' call time. (see criu/cr-restore.c:2663)